### PR TITLE
Land remaining parts of Torchscript Lazy Tensor backend

### PR DESCRIPTION
**Pull request by wconstab**
Differential Revision: D34772403

### CODE REVIEW

**Comment by wconstab** [comment]
cc @antoniojkim - I've been working on this for a while but I think it's pretty close to landing

**Comment by henrytwo** [comment]
Hey @wconstab, thanks for this PR -- I think this will be very helpful for our efforts. Do you have an idea of when this branch will be stable/ready to land in master? I noticed that there are some differences in the codegen compared to what we have in `lazy_tensor_staging`

File: CMakeLists.txt [hunk 1]
```diff
@@ -337,6 +337,9 @@ cmake_dependent_option(USE_CCACHE "Attempt using CCache to wrap the compilation"
 option(WERROR "Build with -Werror supported by the compiler" OFF)
 option(USE_COREML_DELEGATE "Use the CoreML backend through delegate APIs" OFF)
 option(USE_PER_OPERATOR_HEADERS "Whether ATen should generate separate headers for each operator" ON)
+cmake_dependent_option(
```
Comment (wconstab) [inline CMakeLists.txt:340]: @malfet Any chance you can spot what I'm doing wrong here? If you look at failures on this PR's CI, I am observing from cmake.summary that 
INTERN_BUILD_MOBILE: ON
BUILD_LAZY_TS_BACKEND: ON
and I can't figure out why my flag isn't getting turned off under these builds.

File: tools/codegen/gen_lazy_tensor.py [hunk 2]
```diff
@@ -235,12 +240,13 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
...
             "ATen/Operators.h",
+            "ATen/native/CPUFallback.h",
             "torch/csrc/lazy/core/lazy_graph_executor.h",
             "torch/csrc/lazy/core/metrics.h",
             "torch/csrc/lazy/core/shape.h",
-            "lazy_tensor_core/csrc/ts_backend/aten_eager_fallback.h",
-            f"{output_dir}/{backend_key}NativeFunctions.h",
-            f"{output_dir}/{backend_key}LazyIr.h",
+            f"torch/csrc/lazy/generated/{backend_key}NativeFunctions.h",
+            "torch/csrc/lazy/generated/LazyIr.h",
```
Comment (antoniojkim) [inline tools/codegen/gen_lazy_tensor.py:248]: @wconstab Removing this `output_dir` argument effectively forces all LTC backends to use only the generated `LazyIr.h` and `LazyNativeFunctions.h` from the TS backend. This needs to be changes to accept a variable `output_dir` again and set the `output_dir` for the default case to be`torch/csrc/lazy/generated`

**Review by antoniojkim — CHANGES_REQUESTED** [review]

File: tools/codegen/gen_lazy_tensor.py (hunk 2 above)
Comment (wconstab) [reply tools/codegen/gen_lazy_tensor.py:248]: Good catch,  I'm not sure why I did that, maybe there was a reason, but I will try to put it back as an argument.

**Comment by wconstab** [comment]
> Hey @wconstab, thanks for this PR -- I think this will be very helpful for our efforts. Do you have an idea of when this branch will be stable/ready to land in master (e.g. in a few days, or a month)? I noticed that there are some differences in the codegen compared to what we have in `lazy_tensor_staging`, so I want to make sure it's stable before using it

@henrytwo ASAP, in a matter of days hopefully.  I think I just have one minor cmake setting issue now, but one never knows how many more attempts that will take.

File: tools/codegen/gen_lazy_tensor.py [hunk 3]
```diff
@@ -222,9 +223,13 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
                                        grouped_native_functions, backend_key, autograd_key)
 
     # Generate Dispatcher registrations which hook up the nativefunctions
+    internal_headers = [
+        "torch/csrc/lazy/generated/LazyNativeFunctions.h",
```
Comment (henrytwo) [inline tools/codegen/gen_lazy_tensor.py:227]: @wconstab I think a similar change to what was discussed below with the `output_dir` needs to be applied here too, based on my testing

File: tools/codegen/dest/lazy_ir.py [hunk 4]
```diff
@@ -251,7 +251,7 @@ def this_shape(i: int) -> str:
         meta_str += f"""
         TORCH_INTERNAL_ASSERT(shapes.size() == {returns_length});"""
 
-        node_str = f"""auto node = torch::lazy::MakeNode<ir::ops::{schema.node_name}>({node_ctor_input_str},
+        node_str = f"""auto node = torch::lazy::MakeNode<{schema.node_name}>({node_ctor_input_str},
```
Comment (henrytwo) [inline tools/codegen/dest/lazy_ir.py:254]: Why was this changed to exclude the namespace? Should nodes no longer be under the `ir::ops` namespace?

Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:254]: yea, generally we've simplified the namespaces when landing to master.  The ir::ops. namespace was inherited from the torch/xla project.  We use a flatter namespace, almost everything is just under torch::lazy:: now.

**Comment by wconstab** [comment]
Moving to land with several failing CUDA and ASAN unit tests (disabled) and fix separately, because they are also happening on lazy_tensor_staging branch (which only runs CPU mode in CI and no ASAN).

Issues tracked here:
https://github.com/pytorch/pytorch/issues/74519

File: aten/src/ATen/core/op_registration/op_registration_test.cpp [hunk 5]
```diff
@@ -613,14 +614,13 @@ void LazyBackendsAutogradOverridesAutogradKernel(DispatchKey key) {
   EXPECT_FALSE(called_nonautograd);
 }
 
+// no longer test ::Lazy key here
+// since it is now registered to TS backend in-tree and thus behaves differently,
+// does not throw the expected 'could not run..' messages
```
Comment (bdhirsh) [inline aten/src/ATen/core/op_registration/op_registration_test.cpp:619]: I was trying to think of why your changes would affect these tests, since they all use a custom operator (`_test::dummy`). And... it's probably because we now register a CPU fallback kernel to the Lazy key in core? It would probably be better if we just arranged for this test to get built without any of the dispatcher registrations running :p.

Either way, removing seems totally fine.

File: caffe2/CMakeLists.txt [hunk 6]
```diff
@@ -350,18 +350,25 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
       "${TORCH_SRC_DIR}/csrc/autograd/generated/ADInplaceOrViewType_0.cpp"
       "${TORCH_SRC_DIR}/csrc/autograd/generated/ADInplaceOrViewType_1.cpp"
     )
+    if(BUILD_LAZY_TS_BACKEND)
```
Comment (bdhirsh) [inline caffe2/CMakeLists.txt:353]: Mostly just my check for understanding - we're going to unconditionally run the LTC codegen and emit the C++ files, right? And then actually building them looks like it will get guarded by this build flag. (I think this seems fine - it's probably not worth the work to make the codegen conditionally emit LTC code if we don't end up building it).

File: aten/src/ATen/core/op_registration/op_registration_test.cpp (hunk 5 above)
Comment (wconstab) [reply aten/src/ATen/core/op_registration/op_registration_test.cpp:619]: yes, the fallback affects these.

actually, there is another PR out (https://github.com/pytorch/pytorch/pull/73973) which I wanted you to take a look at, and I plan to land after this.  Once that's in, i think we could more easily run tests like this, however, it also doesn't seem like that important of a test to me given the other coverage we'll be adding

File: caffe2/CMakeLists.txt (hunk 6 above)
Comment (wconstab) [reply caffe2/CMakeLists.txt:353]: Right. The codegen runs always, which we could try to avoid i suppose.

File: test/cpp/lazy/test_tensor_impl.cpp [hunk 7]
```diff
@@ -6,12 +6,14 @@
 namespace torch {
 namespace lazy {
 
-// TODO(alanwaketan): Update the following unit tests once the TorchScript backend is merged.
+#ifdef FBCODE_CAFFE2
+// Lazy Tensor is disabled in FBCODE until addressing non-virtual methods (e.g. sizes) in TensorImpl
```
Comment (bdhirsh) [inline test/cpp/lazy/test_tensor_impl.cpp:10]: If this ends up being important, the sizes thing is fixable - @swolchok recently landed something to fix a similar issue for NestedTensor: https://github.com/pytorch/pytorch/pull/73817

File: caffe2/CMakeLists.txt (hunk 6 above)
Comment (bdhirsh) [reply caffe2/CMakeLists.txt:353]: (nah seems fine to me)

File: tools/codegen/dest/register_dispatch_key.py [hunk 8]
```diff
@@ -43,7 +43,9 @@ def gen_registration_headers(
     elif per_operator_headers:
         headers += [
             "#include <ATen/ops/empty.h>",
-            "#include <ATen/ops/empty_strided.h>"]
+            "#include <ATen/ops/empty_strided.h>",
+            "#include <ATen/ops/_copy_from_and_resize.h>",
```
Comment (desertfire) [inline tools/codegen/dest/register_dispatch_key.py:47]: I don't fully understand how `per_operator_headers` works, so just curious why we need to add the extra headers here?

File: tools/codegen/gen_backend_stubs.py [hunk 9]
```diff
@@ -241,15 +241,25 @@ def gen_dispatcher_registrations(
         grouped_native_functions: Sequence[Union[NativeFunction, NativeFunctionsGroup]],
         backend_dispatch_key: DispatchKey,
         dispatch_key: DispatchKey,
-        selector: 'SelectiveBuilder') -> None:
+        selector: 'SelectiveBuilder',
+        build_in_tree: bool = False,
```
Comment (bdhirsh) [inline tools/codegen/gen_backend_stubs.py:245]: Why do we need the flag in `gen_backend_stubs.py` (I thought it wasn't used by LTC?)

Also, I think a light comment here on why we have `build_in_tree` would be nice. Basically, `gen_backend_stubs.py` was originally meant as a codegen pipeline for external backends that don't have the ability to take advantage of `native_functions.yaml` and our core codegen pipeline. And now that LTC is a part of core, it may or may not get built in-tree (depending on whether the backend that hooks into LTC is in-tree or not).

Comment (bdhirsh) [reply tools/codegen/gen_backend_stubs.py:245]: ohh it's just because `gen_lazy_tensor.py` calls into the function in `gen_backend_stubs.py` (so if you're adding a new flag for the LTC codegen, the backend_stubs function needs to know about it). I guess that's fine, since the `gen_backend_stubs.py` entrypoint isn't affected - just the helper functions it uses. (Maybe just include in the comment here that it's only actually needed/used by LTC codegen)

**Review by bdhirsh — APPROVED** [review]
I looked at the codegen, and skimmed the build system changes (but didn't look in detail at the LTC C++ code changes). Looks good!

File: torch/csrc/lazy/ts_backend/ts_backend_impl.cpp [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+  c10::DeviceType c10Type() const {
+    return (c10::DeviceType)type;
+  }
+};
+
+class TSBackendImpl : public torch::lazy::BackendImplInterface {
+ public:
+  TSBackendImpl() : default_device_type_(at::kCPU) {
+    // TODO(whc) unify how all our flags are set and parsed as envs
+    static bool env_use_cuda = std::getenv("LTC_TS_CUDA") != nullptr;
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/csrc/lazy/ts_backend/ts_backend_impl.cpp:31]: yea, it is not ideal.  I copied the way that LTC_.._dynamic shapes mode is set (see ir.cpp)

I think we should change this, but propose to unify them all in a later PR and have one system that lets us override any of the 'FLAGS_' variables using envs with a consistent naming.

File: tools/setup_helpers/generate_code.py [hunk 11]
```diff
@@ -184,24 +189,27 @@ def main() -> None:
     if options.gen_lazy_ts_backend:
         aten_path = os.path.dirname(os.path.dirname(options.native_functions_path))
         ts_backend_yaml = os.path.join(aten_path, 'native/ts_native_functions.yaml')
-
+        ts_native_functions = "torch/csrc/lazy/ts_backend/ts_native_functions.cpp"
```
[the remark this answers is not in the document]
Comment (wconstab) [reply tools/setup_helpers/generate_code.py:192]: > I'll leave it up to you, but the idea with the impl_path argument ...

Wait, i'm confused- It sounds like you think I did not use impl_path and am not opting into extra sanity checking fo ts_native_functions.cpp

I actually think I am opting into the sanity checking and the whole reason I hardcoded the ts_native_functions path here is to enable that.  If I missed something lmk?

File: tools/codegen/dest/register_dispatch_key.py (hunk 8 above)
Comment (bdhirsh) [reply tools/codegen/dest/register_dispatch_key.py:47]: tldr: we used to get away with `#include <ATen/Functions.h>`, which would include every tensor operator (easy for development, but bloats all of our C++ files). The per_operator_headers stuff basically generates a new header file for every aten operator, and enforces that you only include the headers for the operators that you actually need.

The codegen uses the metadata in `native_functions.yaml` to figure out which #includes to stamp out, but there are a few ops that are hardcoded directly into the `RegisterDispatchKey.cpp` template file (that we need to hardcode here, because the codegen would know to include them just based on the yaml).

My guess for why we needed that extra `_copy_from_and_resize` op is because the LTC codegen uses it directly (to handle inplace/out= ops)

File: tools/setup_helpers/generate_code.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (bdhirsh) [reply tools/setup_helpers/generate_code.py:192]: oh nope I just phrased that poorly - you're using it, I just wanted to let you know that it's actually meant to be optional (and gives you some extra error checking) / let you know the history.

File: tools/codegen/gen_backend_stubs.py (hunk 9 above)
Comment (wconstab) [reply tools/codegen/gen_backend_stubs.py:245]: I added a comment locally, but if my CI is clean this time I'm gonna land the comment in a separate diff haha

**Review by desertfire — APPROVED** [review]
LTC C++ part looks good to me.
